From 4d3abb1e4e3abd6d8f5fd62e0937a33029fd94bb Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:19:00 -0400
Subject: [net] netfilter: nft_compat: call {target, match}->destroy() to cleanup entry

Message-id: <20141021161900.22888.8383.stgit@localhost.localdomain>
Patchwork-id: 97970
O-Subject: [RHEL7 net PATCH 072/124] netfilter: nft_compat: call {target, match}->destroy() to cleanup entry
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 3d9b142131ef0cde259dbac5cce36f570fcb4902

 commit 3d9b142131ef0cde259dbac5cce36f570fcb4902
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Wed Jun 11 14:27:46 2014 +0200

    netfilter: nft_compat: call {target, match}->destroy() to cleanup entry

    Otherwise, the reference to external objects (eg. modules) are not
    released when the rules are removed.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nft_compat.c b/net/netfilter/nft_compat.c
index 8a779be..1840989 100644
--- a/net/netfilter/nft_compat.c
+++ b/net/netfilter/nft_compat.c
@@ -195,6 +195,15 @@ static void
 nft_target_destroy(const struct nft_ctx *ctx, const struct nft_expr *expr)
 {
  struct xt_target *target = expr->ops->data;
+ void *info = nft_expr_priv(expr);
+ struct xt_tgdtor_param par;
+
+ par.net = ctx->net;
+ par.target = target;
+ par.targinfo = info;
+ par.family = ctx->afi->family;
+ if (par.target->destroy != NULL)
+  par.target->destroy(&par);
 
  module_put(target->me);
 }
@@ -382,6 +391,15 @@ static void
 nft_match_destroy(const struct nft_ctx *ctx, const struct nft_expr *expr)
 {
  struct xt_match *match = expr->ops->data;
+ void *info = nft_expr_priv(expr);
+ struct xt_mtdtor_param par;
+
+ par.net = ctx->net;
+ par.match = match;
+ par.matchinfo = info;
+ par.family = ctx->afi->family;
+ if (par.match->destroy != NULL)
+  par.match->destroy(&par);
 
  module_put(match->me);
 }
-- 
1.7.1